File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/TypingSignalVectorsTest.kt
package com.flare.im.ui

import java.io.File
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The shared typing table (`spec/typing-vectors.json`), `signal` half: the composer's edits in, the
 * reports this client owes the conversation out. The Vue, Flutter and SwiftUI kits run the same file
 * through the same driver, so a rule that is right here is right there or the difference is a failure.
 */
class TypingSignalVectorsTest {
    @Test fun usesTheTablesConstantsSoTheFourKitsCannotDriftApartQuietly() {
        val rules = (FlareJson.parse(vectorsFile().readText()) as Map<*, *>)["rules"] as Map<*, *>
        assertEquals(FLARE_TYPING_IDLE_STOP_MS, (rules["idleStopMs"] as Number).toInt())
        assertEquals(FLARE_TYPING_REFRESH_MS, (rules["refreshMs"] as Number).toInt())
        assertEquals(FLARE_TYPING_PEER_TTL_MS, (rules["peerTtlMs"] as Number).toInt())
    }

    @Test fun refreshesBeforeThePeersBeliefExpires() {
        // The invariant the table states. Two apps shipped without it and went silent mid-sentence.
        assertTrue(FLARE_TYPING_REFRESH_MS < FLARE_TYPING_PEER_TTL_MS)
    }

    @Test fun everyScriptReportsTheSameThingHere() = runEveryScript(0L)

    /**
     * The same scripts again with the clock where a real one is. A table whose times start at zero is
     * comfortably inside a 32-bit int; `System.currentTimeMillis()` is not, and a rule that stored "now"
     * in one would wrap to a negative instant and never stop typing — a defect no zero-based script can see.
     */
    @Test fun everyScriptReportsTheSameThingWithARealClock() = runEveryScript(1_767_000_000_000L)

    private fun runEveryScript(base: Long) {
        val table = FlareJson.parse(vectorsFile().readText()) as Map<*, *>
        val cases = (table["signal"] as Map<*, *>)["cases"] as List<*>
        assertTrue(cases.size >= 22, "the shared table lost cases: ${cases.size}")
        for (raw in cases) {
            val case = raw as Map<*, *>
            val name = case["name"] as String
            val signal = FlareTypingSignal()
            val reports = mutableListOf<FlareTypingReport>()
            for (rawStep in case["steps"] as List<*>) {
                val step = rawStep as Map<*, *>
                val at = base + (step["at"] as Number).toLong()
                // Time first, then the step: an idle stop that fell due in between must be reported before it.
                reports += signal.tick(at)
                when (step["op"] as String) {
                    "edit" -> reports += signal.edit(step["conversationId"] as? String ?: "", step["text"] as? String ?: "", at)
                    "send" -> reports += signal.send(step["conversationId"] as? String ?: "", at)
                    "close" -> reports += signal.close(at)
                }
            }
            reports += signal.tick(base + (case["endAt"] as Number).toLong())
            val expected = (case["expect"] as List<*>).map {
                val e = it as Map<*, *>
                FlareTypingReport(e["conversationId"] as String, e["typing"] as Boolean, base + (e["at"] as Number).toLong())
            }
            assertEquals(expected, reports, "$name @ base=$base")
        }
    }
}

/** The table lives beside the four kits, so find the repository root from the module's directory. */
private fun vectorsFile(): File {
    var dir: File? = File(".").absoluteFile
    while (dir != null) {
        val candidate = File(dir, "spec/typing-vectors.json")
        if (candidate.isFile) return candidate
        dir = dir.parentFile
    }
    error("spec/typing-vectors.json not found above ${File(".").absolutePath}")
}

File: packages/flutter-im-ui/test/typing_signal_vectors_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter_test/flutter_test.dart';

/// The shared typing table (`spec/typing-vectors.json`), `signal` half: the composer's edits in, the
/// reports this client owes the conversation out. The Vue, SwiftUI and Compose kits run the same file
/// through the same driver, so a rule that is right here is right there or the difference is a failure.
void main() {
  final table = jsonDecode(File('../../spec/typing-vectors.json').readAsStringSync()) as Map<String, dynamic>;
  final rules = table['rules'] as Map<String, dynamic>;
  final cases = ((table['signal'] as Map<String, dynamic>)['cases'] as List).cast<Map<String, dynamic>>();

  test('uses the table\'s constants, so the four kits cannot drift apart quietly', () {
    expect(flareTypingIdleStopMs, rules['idleStopMs']);
    expect(flareTypingRefreshMs, rules['refreshMs']);
    expect(flareTypingPeerTtlMs, rules['peerTtlMs']);
  });

  test('refreshes before the peer\'s belief expires', () {
    // The invariant the table states. Two apps shipped without it and went silent mid-sentence.
    expect(flareTypingRefreshMs, lessThan(flareTypingPeerTtlMs));
  });

  for (final vector in cases) {
    final name = vector['name'] as String;
    // Once at zero and once where a real clock is. A table whose times start at zero is comfortably inside a
    // 32-bit int; a wall clock is not, and a kit that stored "now" in one would wrap to a negative instant
    // and never stop typing — a defect no zero-based script can see. The Compose kit was written that way
    // first; this is the guard that found it.
    for (final base in const [0, 1767000000000]) {
      test('$name${base == 0 ? '' : ' (on a real clock)'}', () {
        final signal = FlareTypingSignal();
        final reports = <FlareTypingReport>[];
        for (final raw in (vector['steps'] as List).cast<Map<String, dynamic>>()) {
          final at = base + (raw['at'] as int);
          // Time first, then the step: an idle stop that fell due in between must be reported before it.
          reports.addAll(signal.tick(at));
          switch (raw['op'] as String) {
            case 'edit':
              reports.addAll(signal.edit(raw['conversationId'] as String? ?? '', raw['text'] as String? ?? '', at));
            case 'send':
              reports.addAll(signal.send(raw['conversationId'] as String? ?? '', at));
            case 'close':
              reports.addAll(signal.close(at));
          }
        }
        reports.addAll(signal.tick(base + (vector['endAt'] as int)));
        final expected = (vector['expect'] as List)
            .cast<Map<String, dynamic>>()
            .map((e) => FlareTypingReport(e['conversationId'] as String, e['typing'] as bool, base + (e['at'] as int)))
            .toList();
        expect(reports, expected, reason: name);
      });
    }
  }
}
